File: backend/calendar_view.py
"""Bounded due-date Calendar projection. Reading never advances Review recurrence."""
from datetime import date, timedelta
from typing import Literal
from fastapi import HTTPException
import review_occurrences

COMMON = ['client_id', 'title', 'status', 'due_date', 'owner_id', 'assignee_id']
REVIEW_FIELDS = ['review_id', 'review_type', 'recurrence', 'custom_recurrence_days', 'schedule_anchor',
                 'current_occurrence_id', 'vendor_purpose', 'period']


def window(start, end):
    try:
        first = date.fromisoformat(start[:10]) if start else date.today() - timedelta(days=42)
        last = date.fromisoformat(end[:10]) if end else first + timedelta(days=92)
        if not 0 <= (last-first).days <= 366:
            raise ValueError()
        return first.isoformat(), last.isoformat()
    except (ValueError, TypeError):
        raise HTTPException(422, 'Choose a valid Calendar period of no more than 367 days.')


def item(row, kind, user, closed, historical=False):
    row = review_occurrences.view(row) if kind == 'review' else row
    terminal = historical or row.get('status') in closed[kind+'s']
    admin = user.get('role') in ('super_admin', 'platform_admin')
    permitted = admin or user.get('role') == 'client_contributor'
    movable = permitted and not terminal and (kind != 'review' or admin and row.get('vendor_purpose') != 'contract')
    oid = row.get('occurrence_id') or review_occurrences.occurrence_id(row) if kind == 'review' else None
    return {'id': row[kind+'_id'], 'key': f'{kind}:{row[kind+"_id"]}:{oid or "current"}',
            'client_id': row['client_id'], 'kind': kind, 'title': row.get('title'), 'status': row.get('status'),
            'owner_id': row.get('owner_id') or row.get('assignee_id'), 'due_date_iso': row.get('due_date'),
            'review_type': row.get('review_type'), 'period': row.get('period') if kind == 'review' else None,
            'occurrence_id': oid, 'historical': terminal, 'can_reschedule': bool(movable)}


async def read(s, query, user, start=None, end=None, scope: Literal['active', 'history', 'all']='active'):
    first, last = window(start, end)
    date_query = {'$gte': first, '$lt': (date.fromisoformat(last)+timedelta(days=1)).isoformat()}
    entries = {}

    def add(row, kind, historical=False):
        value = item(row, kind, user, s.CLOSED, historical)
        if scope == 'active' and value['historical'] or scope == 'history' and not value['historical']:
            return
        entries[value['key']] = value

    for kind in ('review', 'finding', 'task'):
        fields = COMMON + (REVIEW_FIELDS if kind == 'review' else [kind+'_id'])
        status_query = {} if scope == 'all' else {'status': {'$in' if scope == 'history' else '$nin': s.CLOSED[kind+'s']}}
        rows = await s.db[kind+'s'].find({**query, **status_query, 'due_date': date_query}, {'_id': 0, **dict.fromkeys(fields, 1)}).to_list(5001)
        if len(rows) > 5000:
            raise HTTPException(413, 'Too many Calendar entries. Choose a shorter period; no partial results are shown.')
        for row in rows:
            add(row, kind)
    if scope != 'active':
        # Filter and project occurrence snapshots in Mongo; never return whole client history to the browser.
        fields = COMMON + REVIEW_FIELDS + ['occurrence_id']
        history = await s.db.reviews.aggregate([
            {'$match': {**query, 'occurrences': {'$elemMatch': {'due_date': date_query}}}},
            {'$unwind': '$occurrences'}, {'$match': {'occurrences.due_date': date_query}},
            {'$project': {'_id': 0, 'client_id': 1, 'review_id': 1, **dict.fromkeys(['occurrences.'+f for f in fields], 1)}},
            {'$limit': 5001},
        ]).to_list(5001)
        if len(history) > 5000:
            raise HTTPException(413, 'Too many historical Calendar entries. Choose a shorter period.')
        for parent in history:
            occurrence = parent['occurrences']
            if occurrence.get('client_id') not in (None,parent['client_id']) or occurrence.get('review_id') not in (None,parent['review_id']):
                continue
            if occurrence.get('occurrence_id') and occurrence.get('status') in s.CLOSED['reviews']:
                add({**occurrence, 'client_id': parent['client_id'], 'review_id': parent['review_id']}, 'review', True)
    result = {'reviews': {}, 'findings': {}, 'tasks': {}}
    for value in sorted(entries.values(), key=lambda r: (r['due_date_iso'],r['historical'],r['title'] or '',r['key'])):
        result[value['kind']+'s'].setdefault(value['due_date_iso'][:10], []).append(value)
    return result

File: frontend/src/lib/calendarView.js
import rules from './grcRules.json';
import {occurrenceId,reviewView} from './reviewOccurrences';
import {actionStatus} from './actionItems';
import {calendarDay} from './tableFilters';

export const CALENDAR_SCOPES=[['active','Active'],['history','Completed / Closed'],['all','All']];
export const calendarTerminal=(kind,row)=>rules.closed[kind+'s']?.includes(row.status);
export function canMoveCalendar(kind,row,user) {
  const admin=['super_admin','platform_admin'].includes(user?.role);
  return (admin||user?.role==='client_contributor')&&!calendarTerminal(kind,row)&&(kind!=='review'||admin&&row.vendor_purpose!=='contract');
}
export function calendarItem(row,kind,user,historical=false) {
  const source=kind==='review'?reviewView(row):row;
  const oid=kind==='review'?(row.occurrence_id||occurrenceId(row)):null;
  const terminal=historical||calendarTerminal(kind,source);
  return {id:row[kind+'_id'],key:`${kind}:${row[kind+'_id']}:${oid||'current'}`,client_id:row.client_id,kind,title:row.title,status:source.status,
    owner_id:row.owner_id||row.assignee_id||null,due_date_iso:row.due_date,review_type:row.review_type||null,period:kind==='review'?source.period:null,
    occurrence_id:oid,historical:!!terminal,can_reschedule:!terminal&&canMoveCalendar(kind,source,user)};
}
export function calendarWindow(start,end) {
  const first=start?calendarDay(start):calendarDay(new Date().toISOString())-42*86400000;
  const last=end?calendarDay(end):first+92*86400000;
  if(first==null||last==null||last<first||last-first>366*86400000)throw new Error('Choose a valid Calendar period of no more than 367 days.');
  return [first,last];
}
export function calendarBuckets(records,user,{start,end,scope='active'}={}) {
  if(!CALENDAR_SCOPES.some(([key])=>key===scope))throw new Error('Invalid Calendar scope');
  const [first,last]=calendarWindow(start,end),entries=new Map();
  const inRange=row=>{const day=calendarDay(row.due_date);return day!=null&&day>=first&&day<=last;};
  const add=(row,kind,history=false)=>{const value=calendarItem(row,kind,user,history);if(scope==='active'&&value.historical||scope==='history'&&!value.historical)return;entries.set(value.key,value);};
  const bounded=rows=>{if(rows.length>5000)throw new Error('Too many Calendar entries. Choose a shorter period; no partial results are shown.');return rows;};
  for(const kind of ['review','finding','task']) {
    const rows=(records[kind+'s']||[]).filter(r=>inRange(r)&&(scope==='all'||(scope==='history')===!!calendarTerminal(kind,r)));
    for(const row of bounded(rows))add(row,kind);
  }
  if(scope!=='active') {
    const history=(records.reviews||[]).flatMap(r=>(r.occurrences||[]).filter(o=>o.occurrence_id&&calendarTerminal('review',o)&&inRange(o)&&(!o.client_id||o.client_id===r.client_id)&&(!o.review_id||o.review_id===r.review_id)).map(o=>({...o,client_id:r.client_id,review_id:r.review_id})));
    for(const row of bounded(history))add(row,'review',true);
  }
  const result={reviews:{},findings:{},tasks:{}};
  for(const value of [...entries.values()].sort((a,b)=>a.due_date_iso.localeCompare(b.due_date_iso)||Number(a.historical)-Number(b.historical)||(a.title||'').localeCompare(b.title||'')||a.key.localeCompare(b.key))) (result[value.kind+'s'][value.due_date_iso.slice(0,10)]||=[]).push(value);
  return result;
}
export function calendarStatus(item) {
  if(item.kind==='task')return actionStatus(item.status);
  return ({needs_scheduling:'Needs Scheduling',upcoming:'Upcoming',in_progress:'In Progress',completed:'Completed',cancelled:'Cancelled',open:'Open',in_remediation:'In Remediation',remediated:'Pending Validation',closed:'Closed',accepted:'Accepted'})[item.status]||'Status not recorded';
}
export const calendarType=item=>({review:'Review',finding:'Finding',task:'Action Item'})[item.kind]||'Record';
export function calendarSelection(item,record,cid) {
  if(record.client_id!==cid||item.client_id!==cid)throw new Error('Record belongs to another client.');
  if(item.kind!=='review')return {};
  const occurrence=record.occurrences?.find(o=>o.occurrence_id===item.occurrence_id);
  if(occurrence)return {occurrence};
  if(item.occurrence_id!==occurrenceId(record))throw new Error('This Review occurrence is no longer available. Refresh the Calendar.');
  return {};
}
export function rescheduledDate(value,target) {
  if(calendarDay(target)==null)throw new Error('Choose a valid date.');
  // Preserve date-only values, time, offset and precision; this is a date move, not a timezone conversion.
  return target+String(value||'').slice(10);
}
